### PR TITLE
HIVE-27135: AcidUtils#getHdfsDirSnapshots() throws FNFE when a directory is removed in HDFS

### PR DESCRIPTION
**Pull request by mdayakar**
### What changes were proposed in this pull request?
The changes will fix HIVE-27135. Now it will not throw FileNotFoundException even staging directory got removed when AcidUtils#getHdfsDirSnapshots() called.

### Why are the changes needed?
If it throws FileNotFoundException the caller like Initiator/Worker will not get the required details.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added test cases.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java
Comment (SourabhBadhya) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1537]: I am not sure if this function is used by the Cleaner in the current upstream code. AFAIK Cleaner uses `getHdfsDirSnapshotsForCleaner` for getting the map of snapshots containing `HdfsDirSnapshot`.
Link for the function `getHdfsDirSnapshotsForCleaner` - https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java#L1501

On the cleaner side, the function is used here - 
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/handler/CompactionCleaner.java#L269

So from the Cleaner perspective, changing `getHdfsDirSnapshots` might not be very useful.

Comment (mdayakar) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1537]: @SourabhBadhya Thanks for the input. But I could see the other API also getting used here
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/handler/CompactionCleaner.java#L270

The below is the call hierarchy:
image

Comment (SourabhBadhya) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1537]: Yeah the call hierarchy is true, but this is triggered only when the map is `null` in the `getAcidState()` call. In the case of Cleaner, the map of `HdfsDirSnapshots` is already created in prior and is used to call `getAcidState()`.
Relevant code with the null condition -https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java#L1350-L1352

Other usecases of getting the acid state of a table might hit this codepath. But it seems like Cleaner will never hit this codepath.

Comment (mdayakar) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1538]: @SourabhBadhya and @veghlaci05 Now I changed the logic same as _getHdfsDirSnapshotsForCleaner_ API(now all tests are passing). Even the Cleaner is not using this API same issue can happen for others so I corrected it. Please review and merge the code changes.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1541]: I don't really like this approach, you basically moved the logic from the recursive iterator into the addToSnapshot method.

Comment (mdayakar) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1541]: @deniskuzZ I just followed the same approach implemented in _getHdfsDirSnapshotsForCleaner()_ API. Here the main problem with the HADOOP API. Already @ayushtkn raised an issue [HADOOP-18662](https://issues.apache.org/jira/browse/HADOOP-18662) but the fix is not yet merged. If Hadoop fix is merged then no need to do any fix from Hive side.

Comment (veghlaci05) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1541]: @deniskuzZ
I think this approach is fine for now, as @mdayakar mentioned `getHdfsDirSnapshotsForCleaner()` does the same. However, I would create a follow-up task to eliminate the workaround once HADOOP-18662 is merged.

**Review by veghlaci05 — APPROVED** [review]
LGTM, +1

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1541]: 👍 
@mdayakar, could we document the side effects of using `listStatusIterator` & `listFiles` methods (add a javadoc with the link to [HADOOP-18662](https://issues.apache.org/jira/browse/HADOOP-18662))

File: common/src/java/org/apache/hadoop/hive/common/FileUtils.java
Comment (deniskuzZ) [inline common/src/java/org/apache/hadoop/hive/common/FileUtils.java:1379]: why do you need this, you could simply use `listStatusIterator` as in `getHdfsDirSnapshotsForCleaner`

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1552]: why do you need this check if listStatusIterator should already pre-filter the results?

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1549]: are you missing `addToSnapshot(dirToSnapshots, fPath)` here?

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1541]: back to Sourabh's concern, what are we fixing here if that method is not used by Cleaner? cc @SourabhBadhya , @veghlaci05

Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1541]: unless there is another affected place -1 from me

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java
Comment (mdayakar) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1552]: As a part of listStatusIterator we use acidHiddenFileFilter but here we are using acidTempDirFilter which is a different filter.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java
Comment (mdayakar) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1549]: No, `addToSnapshot(dirToSnapshots, fPath) ` need to add if a folder contains a file which is taken care in else part. Same logic exists in the existing code.

File: common/src/java/org/apache/hadoop/hive/common/FileUtils.java
Comment (deniskuzZ) [reply common/src/java/org/apache/hadoop/hive/common/FileUtils.java:1379]: it's not required, FileStatus object is enough

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1552]: 👍

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java
Comment (mdayakar) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java]: Updated accordingly.

File: common/src/java/org/apache/hadoop/hive/common/FileUtils.java
Comment (mdayakar) [reply common/src/java/org/apache/hadoop/hive/common/FileUtils.java]: Changed code to use FileStatus object and using existing _org.apache.hadoop.hive.common.FileUtils#listStatusIterator()_ API which is used in _org.apache.hadoop.hive.ql.io.AcidUtils#getHdfsDirSnapshotsForCleaner()_ API.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java
Comment (mdayakar) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java]: Here it will not list empty directories, actually the above if condition is obsolete in old code. Tested with old and modified code, both don't add empty directories.

Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java]: thanks for checking this!

**Review by deniskuzZ — APPROVED** [review]
LGTM, pending tests

**Pull request merged**
